scipy/optimize/tests/test_optimize.py

```diff
@@ -384,6 +384,20 @@ def test_bfgs_nan(self):
             x = optimize.fmin_bfgs(func, x0, fprime, disp=False)
             assert_(np.isnan(func(x)))
 
+    def test_bfgs_nan_return(self):
+        # Test corner cases where fun returns NaN. See gh-4793.
+
+        # First case: NaN from first call.
+        func = lambda x: np.nan
+        result = optimize.minimize(func, 0)
+        assert_(np.isnan(result['fun']) and result['success'] is False)
```

Comment:
I would have written two asserts, so it's more visible which one fails. (second could be assert_equal